### ISSUE TITLE
Possible to shadow a url from express side?

### ISSUE THREAD

**Issue by adamgajzlerowicz** [szymmis/vite-express#56]
I have a SPA app built with vite. My only problem is that I don't have control over og meta tags when urls are shared on social media.

Is it possible to archive the following with this plugin:
* when user navigates a specific url directly, the ssr would kick in first, rendering template with og tags.
* the SPA routing would take over and user would continue navigating the page never hitting the server rendering again
* if on the other side the user would navigate from another url to that page, then they would only experience ssr.

I feel like that would be a really good solution that would help with SEO, still keeping the SPA architecture, but I can't quite put that together. 

Here is what I tried:

```
import express from 'express'
import ViteExpress from 'vite-express'
import * as fs from "fs";
import path from "path";

const app = express()

app.get('/competitions', async (_, res) => {
    const template = fs.readFileSync(path.resolve( 'index.html'), 'utf-8');
    res.status(200).send(template);
})

ViteExpress.listen(app, 3000, () => console.log('Server is listening...'))
```

But with this I am not using the dist. Not sure if I need to package the html first? 

Thank you

**Comment by szymmis** [comment]
Hi @adamgajzlerowicz, sorry for such a late response but I had quite a busy week 😅.
I think that what you ask is related to #58. If I understand correctly, you want to inject some custom code into the html that's served based on the request parameters? If yes then I guess what was described in #58 would be perfect solution for you.

Basically I would expose some way to define callback that's called inside html serving middleware right before sending it to the client so you'd be able to change the HTML code that's about to be served. You could add some kind of placeholder to the index.html file that's in your project as described in the issue I referenced before, and replace it with the custom code based on the request params.

This callback would have access to the request object, html code after Vite's transformations and would need to return resulting html as a string.

What do you think about this idea?

**Comment by patryk-smc** [comment]
@szymmis thanks for your reply in #58. What you described is more less what I'm doing in my fork, although I pass the data via Express res.locals. Not sure if that's a best practice, but at least I can set my data in the middleware. 

This is what I got:

```ts
export async function injectViteIndexMiddleware(app: core.Express, server: Vite.ViteDevServer) {
  const config = await Vite.resolveConfig({}, "build")
  const template = fs.readFileSync(path.resolve(config.root, "index.html"), "utf8")

  app.get("/*", async (req, res, next) => {
    if (isStaticFilePath(req.path)) {
      next()
    } else {
      const html = await server.transformIndexHtml(req.originalUrl, template)
      const text = injectAppData(html, res.locals.appData)
      res.send(text)
    }
  })
}

export async function injectIndexMiddleware(app: core.Express) {
  const config = await Vite.resolveConfig({}, "build")
  const template = fs.readFileSync(
    path.resolve(config.root, config.build.outDir) + "/index.html",
    "utf8",
  )

  app.use("*", (_, res) => {
    const text = injectAppData(template, res.locals.appData)
    res.send(text)
  })
}
```

**Comment by adamgajzlerowicz** [comment]
Thanks for your response.

I actually found [this article](https://blog.codeminer42.com/making-a-full-stack-app-with-vue-vite-and-express-that-supports-hot-reload/) describing how to do this all from scratch just with vite and express. I was able to get it to work successfully 💃🏼 

I think this is a really cool strategy. It allows to keep existing SPA app, but to inject metadata for bots dynamically, without having to re-implement/rebuild the entire app.

For your plugin I feel like this is a selling point.

**Comment by szymmis** [comment]
Okay so I have a proposal that you can check out in #65. Basically i want to give an option to specify a function that transforms HTML. Just like I was describing before, almost the same thing as you @patryk-smc are doing in your fork. Let me know what you think guys, any feedback always appreciated!

**Comment by szymmis** [comment]
I've released v0.9.0 with these changes. If you are still interested let me know what you think about these changes. Thanks for bringing up this issue @adamgajzlerowicz.
